src/csm_testing/tests/check_workflow_template/__main__.py

```diff
@@ -0,0 +1,314 @@
+):
+    """
+    Find the image in the workflowtemplate and replace its version
+    with the user provided version
+    """
+    updated_workflow_template_str = workflow_template_str.replace(
+        old_image_prefix + old_version, old_image_prefix + new_version
+    )
+    return updated_workflow_template_str
+
+
+def wait_for_workflow_to_succeed(namespace, workflow_name, timeout=6000, interval=20):
+    """
+    Polls the workflow status until it succeeds or the timeout is reached.
+    """
+    start_time = time.time()
+    while time.time() - start_time < timeout:
+        try:
+            # Get the workflow status
+            command = [
+                "kubectl",
+                "get",
+                "workflow",
+                workflow_name,
+                "-n",
+                namespace,
+                "-o",
+                "jsonpath={.status.phase}",
+            ]
+            result = subprocess.run(
+                command,
+                check=True,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                universal_newlines=True,
+            )
+            status = result.stdout.strip()
+
+            # Check the status of the workflow
+            print(f"Workflow {workflow_name} status: {status}")
+
+            if status == "Succeeded":
+                print(f"INFO: Workflow {workflow_name} succeeded.")
+                return True
+
+            if status in ["Failed", "Error"]:
+                print(f"ERROR: Workflow {workflow_name} failed with status {status}.")
+                return False
+
+            if status == "Running":
+                print("INFO: Check if main is done")
+                if check_and_kill_pod(workflow_name, namespace):
+                    print(
+                        f"WARNING: {workflow_name} is running but the main container is done."
+                    )
+                    return True
+        except ApiException as err:
+            print(f"ERROR: Exception when fetching workflow status: {err}")
+            return False
+
+        # Wait before polling again
+        time.sleep(interval)
+
+    print(f"ERROR: Workflow {workflow_name} did not complete within {timeout} seconds.")
+    return False
+
+
+def delete_resources(namespace, workflow_name, workflow_template_name):
+    """Delete the workflow, workflow template, and associated pods."""
+
+    for resource, name in [
+        ("workflow", workflow_name),
+        ("workflowtemplate", workflow_template_name),
+    ]:
+        try:
+            print(f"INFO: Deleting {resource} {name}...")
+            subprocess.run(
+                ["kubectl", "delete", resource, name, "-n", namespace], check=True
+            )
+            print(f"INFO: {resource.capitalize()} {name} deleted successfully.")
+        except subprocess.CalledProcessError as err:
+            print(f"ERROR: Failed to delete {resource} {name}: {err}")
```

Comment:
But it seems like the test will still pass, in that case, even though there was a problem. You can still have the cleanup cover the other resources and then have the test end up failing. For example, you could modify the Goss YAML test files so that they fail if they see 'ERROR' in the output of the test.

Reply:
we don't want to test case to fail in case the cleanup is not successful. So rather than keeping the log as ERROR statements, I have made them as WARNING.

Reply:
This is fine as long as the resources left behind aren't things that will cause problems for anyone else using the system (or cause future iterations of the test to fail).